### ISSUE TITLE
Postgres support for ERPNext

### ISSUE THREAD

**Issue by rmehta** [frappe/erpnext#24389]
Frappe Framework has support for both MariaDB and Postgres. ERPNext does not have support for Postgres. While queries are mostly similar, there are 1300 SQL queries that need to be verified and checked. Thankfully we have unit-tests.

Steps:

1. Setup tests for postgres (similar to Frappe Framework)
2. Makes sure the tests pass for all queries

If there are problems in certain SQL queries:

1. Use the ORM
2. Rewrite the query so that it will work for both databases
3. Use `frappe.db.multi_sql` to write separate queries for each platform (for complex joins)

**Comment by kisg** [comment]
Is someone at Frappe already working on this? We would be interested in contributing code to this issue.

I submitted a proposal some time ago on the forum, that also included more proposed data layer refactoring at the same time: https://discuss.erpnext.com/t/erpnext-postgres-support/67069/2

Would there be interest in such changes? 

It does not have to be an "all or nothing" acceptance, e.g. I can understand if the numeric id part is kept as a completely separate project, because it would require creating prototypes first and do extensive benchmarking to make sure that the transition is worth it. 

But the following changes certainly would bring a lot of maintainability and performance benefits:
- Move to a DAO model where the SQL queries are kept in separate files in an app.
- Introduce a Fluent SQL query API that can hide most of the syntax differences across SQL engines. For example Pypika (https://github.com/kayak/pypika) could be used as a starting point.
- Use prepared statements in the ORM and in custom SQL queries as well

If there is interest, I would be happy to discuss the technical details and provide a more detailed design for review either here or on the forum.

**Comment by rmehta** [comment]
@kisg no we have not yes assigned this. If you are working on the first milestone of setting up the test and making sure they pass with postgres, do open a PR - will get someone from the Frappe team to help you.

**Comment by pratikbin** [comment]
waiting for this one

**Comment by Verequies** [comment]
How is the progress on this? Any sort of timeline? Would love to see this integrated.

**Comment by elbanna-h** [comment]
Yes, Postgres is more stable and powerful

**Comment by Verequies** [comment]
I'm actually working on fixing the queries at the moment. Currently the installation works and am surprised that a lot of it actually works as is.

**Comment by wodka** [comment]
@Verequies is there any PR / branch others can help you to get it running?

**Comment by neilcamilleri** [comment]
Also waiting for this, this is a blocker to adoption for us.

**Comment by Verequies** [comment]
Sorry for the lack of updates, I plan to get back to this however as I just graduated from university I am stuck on finishing a couple of personal projects first - at least to the point where I can release an initial version. I am almost finished however.

When I get back to it, I am happy for others to help :) @wodka

**Comment by JeffKnorr** [comment]
@Verequies I'd be interested in helping too.  Please share your progress so far.
> 
> 
> Sorry for the lack of updates, I plan to get back to this however as I just graduated from university I am stuck on finishing a couple of personal projects first - at least to the point where I can release an initial version. I am almost finished however.
> 
> When I get back to it, I am happy for others to help :) @wodka

**Comment by c-p-b** [comment]
Frappe: https://github.com/frappe/frappe/pull/14963
ERPNext: https://github.com/frappe/erpnext/pull/28359

**Comment by Verequies** [comment]
@cpdeethree Thanks for taking over, glad to see that some of my commits helped. I wasn't sure when I was going to be able to get back to due to running my own businesses and doing contract jobs so glad to see it is being worked on.

**Comment by iashaik** [comment]
Is this still active? It's been months.

**Comment by skygunner** [comment]
This issue might be related 
https://github.com/frappe/frappe/issues/19996

**Comment by sidharthramesh** [comment]
Hey @ankush, it seems like we are focusing on long-term maintainability as mentioned here https://github.com/frappe/erpnext/pull/28359#issuecomment-1277105369, which is great!

I know it's hard to estimate timelines for an open-source project, however, is there any information or idea you have on how long it would be before ERPNext can run on Postgres? 

We need to make a few business decisions around this since the rest of our stack already uses Postgres and running an instance of MariaDB might increase our cloud costs significantly.

CC: @pythonpen @DanBrown47

**Comment by rmehta** [comment]
@sidharthramesh if someone out there is willing to contribute (there is a semi-finished PR out there), then we would be happy to accept as long as it meets quality and contribution guidelines

**Comment by ankush** [comment]
> I know it's hard to estimate timelines for an open-source project, however, is there any information or idea you have on how long it would be before ERPNext can run on Postgres?

If someone works continuously, my guess would be approximately 3-9 months. That will be enough to reach a level where it can be used in production. 

I get the cost of running a separate server but Frappe on MariaDB is battle tested for years and optimized over the years, all of that will take at least a year of running postgres in prod to reach a similar level of maturity. So don't forget that intangible cost. 



---

I'll at least start putting all the work that is required in issue so whoever ends up working on it (frappe or outside) can have some structure way of addressing the problem.

**Comment by kisg** [comment]
One of the issues with adding support for PostgreSQL is the continued maintenance of the sql() and multi_sql() queries. What is your position on moving away from these in favor of the new PyPika-based Query Builder interface? I would even go as far as deprecating the old sql() and multi_sql() APIs now, before ERPNext 15 is released, and maybe remove it in the ERPNext 16 cycle. (I know that this is a drastic move, and would require agreement from both the core developers and the community.)

What do you think @ankush?

**Comment by sidharthramesh** [comment]
> > I know it's hard to estimate timelines for an open-source project, however, is there any information or idea you have on how long it would be before ERPNext can run on Postgres?
> 
> If someone works continuously, my guess would be approximately 3-9 months. That will be enough to reach a level where it can be used in production.
> 
> I get the cost of running a separate server but Frappe on MariaDB is battle tested for years and optimized over the years, all of that will take at least a year of running postgres in prod to reach a similar level of maturity. So forget that that intangible cost.
> 
> I'll at least start putting all the work that is required in issue so whoever ends up working on it (frappe or outside) can have some structure way of addressing the problem.

Thank you @ankush for the detailed reply. I understand the implications and the fact that Postgres support will never be on par with MariaDB.

However, assuming that we (and other interested parties) have commercial interest in getting production ready Postgres support, would it be useful to discuss methods to fund a work group / developer / ERPNext team for this in the long term? Are there any governance structure already in place for ways to fund feature requests like this?

**Comment by ankush** [comment]
I've started listing down everything required for stable support in framework here: https://github.com/frappe/frappe/issues/21613 (keep in mind this will keep expanding as we find more issues)


---

> However, assuming that we (and other interested parties) have commercial interest in getting production ready Postgres support, would it be useful to discuss methods to fund a work group / developer / ERPNext team for this in the long term? Are there any governance structure already in place for ways to fund feature requests like this?

As of now I need an internal owner first to push postgres support. I'll get back to you after that.  

I can push for stable support for Framework, ERPNext support is mostly rewriting queries in QB or ORM OR using `multisql`. Community developers can help by sending small PRs that move raw queries to QB/ORM. Make sure it's tested and old/new generated queries are equivalent. We don't want to introduce new bugs in process of rewriting queries. 

If you're interested we can build a small working group on telegram and collaborate on it. Ping me on telegram: `@ankushmnt` or email `ankush (at) frappe.io`

---

> One of the issues with adding support for PostgreSQL is the continued maintenance of the sql() and multi_sql() queries. What is your position on moving away from these in favor of the new PyPika-based Query Builder interface? I would even go as far as deprecating the old sql() and multi_sql() APIs now, before ERPNext 15 is released, and maybe remove it in the ERPNext 16 cycle. (I know that this is a drastic move, and would require agreement from both the core developers and the community.)

We discourage use of `sql` already. Removing `sql` would be a bit too much. Apps who want to support postgres can chose to not use it by using https://github.com/frappe/semgrep-rules

**Comment by ankush** [comment]
Link to telegram group if anyone wants to participate `t.me/    +cPr4H4qhiOJlNjc1` (remove spaces)

**Comment by pramod-shanmugam** [comment]
What are the steps that you followed to setup postgres instead on mariadb?

**Comment by SalahAdDin** [comment]
Is there any ETA for this feature?

**Comment by WilsonG08** [comment]
Alguien me puede ayudar, estoy utilizando ERPNext, frappe v=14 y postgress pero me da este error:

2024-05-02 12:51:29 2024-05-02 17:51:29.147 UTC [372] ERROR:  syntax error at or near "index" at character 92
2024-05-02 12:51:29 2024-05-02 17:51:29.147 UTC [372] STATEMENT:  select item_code, warehouse, count(name) as total_row
2024-05-02 12:51:29                     from "tabStock Ledger Entry" force index (item_warehouse)
2024-05-02 12:51:29                     where
2024-05-02 12:51:29                             (warehouse = 'Stores - W'
2024-05-02 12:51:29                                     and item_code in ('ITM-0001'))
2024-05-02 12:51:29                             and timestamp(posting_date, posting_time)
2024-05-02 12:51:29                                     >= timestamp('2024-05-02'::date, '12:51:24.913741')
2024-05-02 12:51:29                             and voucher_no != 'MAT-STE-2024-00001'
2024-05-02 12:51:29                             and is_cancelled = '0'
2024-05-02 12:51:29                     GROUP BY
2024-05-02 12:51:29                             item_code, warehouse
2024-05-02 12:54:07 2024-05-02 17:54:07.177 UTC [389] ERROR:  syntax error at or near "index" at character 92
2024-05-02 12:54:07 2024-05-02 17:54:07.177 UTC [389] STATEMENT:  select item_code, warehouse, count(name) as total_row
2024-05-02 12:54:07                     from "tabStock Ledger Entry" force index (item_warehouse)
2024-05-02 12:54:07                     where
2024-05-02 12:54:07                             (warehouse = 'Stores - W'
2024-05-02 12:54:07                                     and item_code in ('ITM-0001'))
2024-05-02 12:54:07                             and timestamp(posting_date, posting_time)
2024-05-02 12:54:07                                     >= timestamp('2024-05-02'::date, '12:54:06.508791')
2024-05-02 12:54:07                             and voucher_no != 'MAT-STE-2024-00001'
2024-05-02 12:54:07                             and is_cancelled = '0'
2024-05-02 12:54:07                     GROUP BY
2024-05-02 12:54:07                             item_code, warehouse

**Comment by clayrisser** [comment]
I'm really looking forward to erpnext postgres support.

**Comment by Amein-Eskinder** [comment]
Same here, Signing up for future notification

**Comment by psyberfly** [comment]
Hello!
What is the status of ERP-Next and postgres compatibility? 
I need to host a postgres compatible ERP for an e-commerce business, is the integration possible and if so production ready?

**Comment by Amein-Eskinder** [comment]
https://gitpay.me/#/task/984

If you want to contribute with a decent bounty you can start by following the link

**Comment by omarsoufiane** [comment]
Would creating a bounty on this issue help solve it faster?

**Comment by testn** [comment]
Is there a list of todo items? I can help work on this

**Comment by C-L-STARK** [comment]
#46000

**Comment by tctlrd** [comment]
This might be relevant to the conversation:
https://www.amazingcto.com/postgres-for-everything/

![JustPostgres](https://www.amazingcto.com/images/JustPostgres.png.webp)

**Comment by surfmuggle** [comment]
@tctlrd First of thanks for the link. I agree that in many projects a simpler stack would be good enough ( [Consider sqlite](https://blog.wesleyac.com/posts/consider-sqlite) from Wesley Aptekar-Cassels is worth reading). 

Could it be that at least one link from the article above requires any additonal steps?
From this

> Use Postgres as a message queue with [SKIP LOCKED](https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/) instead of Kafka (if you only need a message queue).

the link to [What is SKIP LOCKED for in PostgreSQL 9.5?](https://www.enterprisedb.com/blog/what-skip-locked-postgresql-95-0) does not show any text for me.


![Image](https://github.com/user-attachments/assets/6656a785-6737-4a30-a0c5-0f2ef8a89321)

**Comment by Ajaay7** [comment]
can we expect any support of postgres in version 16

**Comment by shivekkhurana** [comment]
Related: https://discuss.frappe.io/t/using-postgresql-as-a-database-for-erpnext/1297/55

**Comment by SalahAdDin** [comment]
> Related: https://discuss.frappe.io/t/using-postgresql-as-a-database-for-erpnext/1297/55

It looks they won't work on this.

**Comment by surfmuggle** [comment]
> > Related: https://discuss.frappe.io/t/using-postgresql-as-a-database-for-erpnext/1297/55
> 
> It looks they won't work on this.

The discussion was about forking this repo. After a 10-second search, I found https://github.com/VikashMediboina/frappe-crm-postgres No idea if this is just a random fork. 
@brian-pond Is there another fork to support Postgres?

**Comment by SalahAdDin** [comment]
> > > Related: https://discuss.frappe.io/t/using-postgresql-as-a-database-for-erpnext/1297/55
> > 
> > 
> > It looks they won't work on this.
> 
> The discussion was about forking this repo. After a 10 second search i found https://github.com/VikashMediboina/frappe-crm-postgres - no idea if this is just a random fork. [@brian-pond](https://github.com/brian-pond) Is there another fork to support Postgres?

It should not be necessary to fork the system, just to modularizate it.

**Comment by brian-pond** [comment]
To answer your question, @surfmuggle:

TL;DR - There are no forks that are _clean, stable, and ready for use_.

#### Context
Last year I helped an ERPNext client switch to Postgres.  It's important to note that:
1. I performed that work in an _already-forked_ repo. 
2. I was allotted very little time to achieve this.  And I never intended others to use my solution.

Because I was in a hurry, I took -many- shortcuts.  I did many improper things.  I broke MariaDB compatibility in favor of Postgres.

My experiment _was_ very successful.  PostgreSQL immediately boosted performance, reduced lock contention, and provided a stable Replication system.  It was a huge win.

But my fork is definitely -not- usable by others.  It's a messy proof-of-concept.

A good reusable solution requires a lot more (and better) work that I performed.

The best path forward is for Frappe and ERPNext Maintainers to agree that Postgres should be supported, and begin accepting PR's from the community.  Maintaining separate forks isn't sustainable.

CC:  @agritheory

**Comment by tctlrd** [comment]
@brian-pond Encouraging to hear of your report that switching ERPNext from MariaDB to PostgreSQL improved performance, reduced lock contention, and provided a stable replication system. Thank you for sharing!

> My experiment was very successful. PostgreSQL immediately boosted performance, reduced lock contention, and provided a stable Replication system. It was a huge win.

**Comment by pythonpen** [comment]
@brian-pond :
Good to know about efforts and validations - it sure now justifies Postgres support for the main repo and the community to rally towards it.

**Comment by davispuh** [comment]
I just tried installing latest EPRNext v16.11.0 with PostgreSQL and it installed fine and it "appears" to "work" but seems there are bunch of SQL queries that are not portable and need fixing.
Some errors I noticed
```
   File "/usr/local/lib/python3.14/threading.py", line 1044, in _bootstrap
     self._bootstrap_inner()
   File "/usr/local/lib/python3.14/threading.py", line 1082, in _bootstrap_inner
     self._context.run(self.run)
   File "/usr/local/lib/python3.14/threading.py", line 1024, in run
     self._target(*self._args, **self._kwargs)
   File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 119, in _worker
     work_item.run(ctx)
   File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 86, in run
     result = ctx.run(self.task)
   File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 73, in run
     return fn(*args, **kwargs)
   File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 300, in handle
     keepalive = self.handle_request(req, conn)
   File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 352, in handle_request
     respiter = self.wsgi(environ, resp.start_response)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 84, in application
     app(environ, start_response),
   File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/werkzeug/wrappers/request.py", line 193, in application
     resp = f(*args[:-2] + (request,))
   File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 121, in application
     response = frappe.api.handle(request)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/api/__init__.py", line 63, in handle
     data = endpoint(**arguments)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/api/v1.py", line 40, in handle_rpc_call
     return frappe.handler.handle()
   File "/home/frappe/frappe-bench/apps/frappe/frappe/handler.py", line 53, in handle
     data = execute_cmd(cmd)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/handler.py", line 86, in execute_cmd
     return frappe.call(method, **frappe.form_dict)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 1129, in call
     return fn(*args, **newargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/typing_validations.py", line 36, in wrapper
     return func(*args, **kwargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/desk/doctype/number_card/number_card.py", line 152, in get_result
     res = frappe.get_list(
   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 1362, in get_list
     return frappe.model.qb_query.DatabaseQuery(doctype).execute(*args, **kwargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/model/qb_query.py", line 212, in execute
     result = query.run(debug=debug, as_dict=not as_list, update=update)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/query_builder/utils.py", line 131, in execute_query
     result = frappe.local.db.sql(query, params, *args, **kwargs)  # nosemgrep
   File "/home/frappe/frappe-bench/apps/frappe/frappe/database/postgres/database.py", line 234, in sql
     return super().sql(modify_query(query), modify_values(values), *args, **kwargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 294, in sql
     traceback.print_stack()
 Error in query:
 column "tabPurchase Invoice.creation" must appear in the GROUP BY clause or be used in an aggregate function
 LINE 1: ...01' AND '2026-12-31' AND "company"='aaa' ORDER BY "creation"...
                                                              ^
 
   File "/usr/local/lib/python3.14/threading.py", line 1044, in _bootstrap
     self._bootstrap_inner()
   File "/usr/local/lib/python3.14/threading.py", line 1082, in _bootstrap_inner
     self._context.run(self.run)
   File "/usr/local/lib/python3.14/threading.py", line 1024, in run
     self._target(*self._args, **self._kwargs)
   File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 119, in _worker
     work_item.run(ctx)
   File "/usr/local/lib/python3.14/threading.py", line 1044, in _bootstrap
     self._bootstrap_inner()
   File "/usr/local/lib/python3.14/threading.py", line 1082, in _bootstrap_inner
     self._context.run(self.run)
   File "/usr/local/lib/python3.14/threading.py", line 1024, in run
     self._target(*self._args, **self._kwargs)
   File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 86, in run
     result = ctx.run(self.task)
   File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 73, in run
     return fn(*args, **kwargs)
   File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 119, in _worker
     work_item.run(ctx)
   File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 86, in run
     result = ctx.run(self.task)
   File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 73, in run
     return fn(*args, **kwargs)
   File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 300, in handle
     keepalive = self.handle_request(req, conn)
   File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 352, in handle_request
     respiter = self.wsgi(environ, resp.start_response)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 84, in application
     app(environ, start_response),
   File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/werkzeug/wrappers/request.py", line 193, in application
     resp = f(*args[:-2] + (request,))
   File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 121, in application
     response = frappe.api.handle(request)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/api/__init__.py", line 63, in handle
     data = endpoint(**arguments)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/api/v1.py", line 40, in handle_rpc_call
     return frappe.handler.handle()
   File "/home/frappe/frappe-bench/apps/frappe/frappe/handler.py", line 53, in handle
     data = execute_cmd(cmd)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/handler.py", line 86, in execute_cmd
     return frappe.call(method, **frappe.form_dict)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 1129, in call
     return fn(*args, **newargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/typing_validations.py", line 36, in wrapper
     return func(*args, **kwargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/desk/doctype/number_card/number_card.py", line 152, in get_result
     res = frappe.get_list(
   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 1362, in get_list
     return frappe.model.qb_query.DatabaseQuery(doctype).execute(*args, **kwargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/model/qb_query.py", line 212, in execute
     result = query.run(debug=debug, as_dict=not as_list, update=update)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/query_builder/utils.py", line 131, in execute_query
     result = frappe.local.db.sql(query, params, *args, **kwargs)  # nosemgrep
   File "/home/frappe/frappe-bench/apps/frappe/frappe/database/postgres/database.py", line 234, in sql
     return super().sql(modify_query(query), modify_values(values), *args, **kwargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 294, in sql
     traceback.print_stack()
 Error in query:
 column "tabPayment Entry.creation" must appear in the GROUP BY clause or be used in an aggregate function
 LINE 1: ...ment_type"='Receive' AND "company"='aaa' ORDER BY "creation"...
                                                              ^

   File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 300, in handle
     keepalive = self.handle_request(req, conn)
   File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 352, in handle_request
     respiter = self.wsgi(environ, resp.start_response)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 84, in application
     app(environ, start_response),
   File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/werkzeug/wrappers/request.py", line 193, in application
     resp = f(*args[:-2] + (request,))
   File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 121, in application
     response = frappe.api.handle(request)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/api/__init__.py", line 63, in handle
     data = endpoint(**arguments)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/api/v1.py", line 40, in handle_rpc_call
     return frappe.handler.handle()
   File "/home/frappe/frappe-bench/apps/frappe/frappe/handler.py", line 53, in handle
     data = execute_cmd(cmd)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/handler.py", line 86, in execute_cmd
     return frappe.call(method, **frappe.form_dict)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 1129, in call
     return fn(*args, **newargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/typing_validations.py", line 36, in wrapper
     return func(*args, **kwargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/desk/doctype/number_card/number_card.py", line 152, in get_result
     res = frappe.get_list(
   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 1362, in get_list
     return frappe.model.qb_query.DatabaseQuery(doctype).execute(*args, **kwargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/model/qb_query.py", line 212, in execute
     result = query.run(debug=debug, as_dict=not as_list, update=update)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/query_builder/utils.py", line 131, in execute_query
     result = frappe.local.db.sql(query, params, *args, **kwargs)  # nosemgrep
   File "/home/frappe/frappe-bench/apps/frappe/frappe/database/postgres/database.py", line 234, in sql
     return super().sql(modify_query(query), modify_values(values), *args, **kwargs)
   File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 294, in sql
     traceback.print_stack()
 Error in query:
 column "tabSales Invoice.creation" must appear in the GROUP BY clause or be used in an aggregate function
 LINE 1: ...01' AND '2026-12-31' AND "company"='aaa' ORDER BY "creation"...

Syntax error in query:
SELECT "account","debit","credit","debit_in_account_currency","credit_in_account_currency","account_currency","posting_date","is_opening","fiscal_year" FROM "tabGL Entry" FORCE INDEX ("posting_date_company_index") WHERE "company"=%(param1)s AND "is_cancelled"= '0' AND "posting_date"<='2026-12-31' AND "voucher_type"<>%(param2)s AND "posting_date">='2026-01-01' AND ("finance_book" IN (%(param3)s,%(param4)s,%(param5)s) OR "finance_book" IS NULL) AND EXISTS (SELECT "tabAccount"."name" FROM "tabAccount" WHERE "tabAccount"."name"="tabGL Entry"."account" AND "tabAccount"."is_group"= '0' AND "tabAccount"."lft">= '137' AND "tabAccount"."rgt"<= '150' AND "tabAccount"."root_type"=%(param6)s) {'param1': 'aaa', 'param2': 'Period Closing Voucher', 'param3': '', 'param4': '', 'param5': '', 'param6': 'Income'}
  File "/usr/local/lib/python3.14/threading.py", line 1044, in _bootstrap
    self._bootstrap_inner()
  File "/usr/local/lib/python3.14/threading.py", line 1082, in _bootstrap_inner
    self._context.run(self.run)
  File "/usr/local/lib/python3.14/threading.py", line 1024, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 119, in _worker
    work_item.run(ctx)
  File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 86, in run
    result = ctx.run(self.task)
  File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 73, in run
    return fn(*args, **kwargs)
  File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 300, in handle
    keepalive = self.handle_request(req, conn)
  File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 352, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 84, in application
    app(environ, start_response),
  File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/werkzeug/wrappers/request.py", line 193, in application
    resp = f(*args[:-2] + (request,))
  File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 121, in application
    response = frappe.api.handle(request)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/api/__init__.py", line 63, in handle
    data = endpoint(**arguments)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/api/v1.py", line 40, in handle_rpc_call
    return frappe.handler.handle()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/handler.py", line 53, in handle
    data = execute_cmd(cmd)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/handler.py", line 86, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 1129, in call
    return fn(*args, **newargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/typing_validations.py", line 36, in wrapper
    return func(*args, **kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 491, in wrapper_fn
    retval = fn(*args, **get_newargs(fn, kwargs))
  File "/home/frappe/frappe-bench/apps/frappe/frappe/desk/query_report.py", line 234, in run
    frappe.log_error("Report Error")
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/error.py", line 64, in log_error
    error_log = frappe.get_doc(
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/utils/__init__.py", line 217, in wrapper
    return func(*args, **kw)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/document.py", line 111, in get_doc
    return get_doc_from_dict(kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/document.py", line 140, in get_doc_from_dict
    if controller := get_controller(data["doctype"]):
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/base_document.py", line 92, in get_controller
    site_controllers[doctype] = import_controller(doctype)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/base_document.py", line 103, in import_controller
    doctype_info = frappe.db.get_value("DocType", doctype, ("module", "custom", "is_tree"), as_dict=True)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 561, in get_value
    result = self.get_values(
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 677, in get_values
    out = query.run(as_dict=as_dict, debug=debug, update=update, run=run, pluck=pluck)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/query_builder/utils.py", line 131, in execute_query
    result = frappe.local.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/postgres/database.py", line 234, in sql
    return super().sql(modify_query(query), modify_values(values), *args, **kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 294, in sql
    traceback.print_stack()
Error in query:
current transaction is aborted, commands ignored until end of transaction block

  File "/usr/local/lib/python3.14/threading.py", line 1044, in _bootstrap
    self._bootstrap_inner()
  File "/usr/local/lib/python3.14/threading.py", line 1082, in _bootstrap_inner
    self._context.run(self.run)
  File "/usr/local/lib/python3.14/threading.py", line 1024, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 119, in _worker
    work_item.run(ctx)
  File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 86, in run
    result = ctx.run(self.task)
  File "/usr/local/lib/python3.14/concurrent/futures/thread.py", line 73, in run
    return fn(*args, **kwargs)
  File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 300, in handle
    keepalive = self.handle_request(req, conn)
  File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/gunicorn/workers/gthread.py", line 352, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 84, in application
    app(environ, start_response),
  File "/home/frappe/frappe-bench/env/lib/python3.14/site-packages/werkzeug/wrappers/request.py", line 193, in application
    resp = f(*args[:-2] + (request,))
  File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 139, in application
    response = e.get_response(request.environ) if isinstance(e, HTTPException) else handle_exception(e)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/permissions.py", line 930, in wrapper
    return fn(e, *args, **kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/app.py", line 407, in handle_exception
    log_error_snapshot(e)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/error.py", line 125, in log_error_snapshot
    log_error(title=str(exception), defer_insert=True)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/error.py", line 64, in log_error
    error_log = frappe.get_doc(
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/utils/__init__.py", line 217, in wrapper
    return func(*args, **kw)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/document.py", line 111, in get_doc
    return get_doc_from_dict(kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/document.py", line 140, in get_doc_from_dict
    if controller := get_controller(data["doctype"]):
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/base_document.py", line 92, in get_controller
    site_controllers[doctype] = import_controller(doctype)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/model/base_document.py", line 103, in import_controller
    doctype_info = frappe.db.get_value("DocType", doctype, ("module", "custom", "is_tree"), as_dict=True)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 561, in get_value
    result = self.get_values(
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 677, in get_values
    out = query.run(as_dict=as_dict, debug=debug, update=update, run=run, pluck=pluck)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/query_builder/utils.py", line 131, in execute_query
    result = frappe.local.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/postgres/database.py", line 234, in sql
    return super().sql(modify_query(query), modify_values(values), *args, **kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 294, in sql
    traceback.print_stack()
Error in query:
current transaction is aborted, commands ignored until end of transaction block
```

So I think it's doable to fix these issues one by one and get it working.
Someone just have to test all features and fix all of these broken queries...

**Comment by agritheory** [comment]
@davispuh We've written a codemod that does most of the that. I recommend a flow like:
 - rewrite the queries in the tests in scope
 - run the tests in scope with MariaDB
 - rewrite the code under test
 - run the tests in scope with both MariaDB and Postgres
 - rinse and repeat

https://github.com/agritheory/test_utils/blob/main/docs/sql_registry.md

**Comment by mateu** [comment]
Thanks, I was looking to see if ERPNext support PostgreSQL, and found the answer is a fresh yes!
